### PR TITLE
Auto-generated conversion methods calling one another

### PR DESCRIPTION
**Pull request by wojtek-t**
Part of #6800 

Conversion time decreased by:

```
benchmark                                    old ns/op     new ns/op     delta
BenchmarkPodConversion                       149268        30690         -79.44%
BenchmarkNodeConversion                      60733         35517         -41.52%
BenchmarkReplicationControllerConversion     135528        32236         -76.21%
```

Comparing with no conversion methods:

```
benchmark                                    old ns/op     new ns/op     delta
BenchmarkPodConversion                       388502        30690         -92.10%
BenchmarkNodeConversion                      192973        35517         -81.59%
BenchmarkReplicationControllerConversion     385409        32236         -91.64%
```

@smarterclayton @lavalamp - your predictions of 10X improvements were so close to what it gives :)

cc @smarterclayton @lavalamp @fgrzadkowski

### CODE REVIEW

**Comment by wojtek-t** [comment]
Note: Please do NOT merge #7560 and this PR at the same time - they need to be rebased before merging the other one.

**Comment by smarterclayton** [comment]
Woohoo!  What's the memory use look like?

----- Original Message -----

**Comment by wojtek-t** [comment]
>
Woohoo!  What's the memory use look like?

When I checked it in the morning the gain was significant but I can't remember it now. I will rerun the experiments tomorrow (I need to run now).

BTW PR rebased

**Comment by smarterclayton** [comment]
Can we move the generated functions to their own file conversion_generated.go?

**Comment by lavalamp** [comment]
Awesome-- +1 for moving to a new file. I'll give a more detailed look later.

**Comment by wojtek-t** [comment]
>
Can we move the generated functions to their own file conversion_generated.go?

I agree that's a good idea. However, since today is a public holiday in Poland I might not find time to do it before Monday.

**Comment by wojtek-t** [comment]
Regarding memory, here is the comparison between having no generated conversions and this PR id ~3.5X better - the graphs below:

Before:
profile_old

After:
profile_new

Also, I've updated PR so that the generated code is in a separate file.

PTAL

**Comment by wojtek-t** [comment]
BTW, once we agree that this is the final format, I will generate conversions to v1 api version in a separate PR.

**Comment by wojtek-t** [comment]
**Update on memory**
I think the the gain is bigger than 3.5X that i mentioned. Basically without generated conversion functions, there are 900 repetitions of methods, and with them generated 14.000 repetitions. So I think that we should multiply this 3.5X by 15 :) [or sth like that].

**Comment by wojtek-t** [comment]
@lavalamp can you please take a look into it later today? Thanks!

**Comment by lavalamp** [comment]
yes I will get to this by EOD

File: pkg/conversion/generator.go [hunk 1]
```diff
@@ -458,12 +481,32 @@ func (g *generator) writeConversionForPtr(w io.Writer, inField, outField reflect
 		}
 	}
 
-	assignFormat := "if err := s.Convert(&in.%s, &out.%s, 0); err != nil {\n"
-	assignStmt := fmt.Sprintf(assignFormat, inField.Name, outField.Name)
-	if err := writeLine(w, indent, assignStmt); err != nil {
+	ifFormat := "if in.%s != nil {\n"
```
Comment (lavalamp) [inline pkg/conversion/generator.go]: I'm a little nervous that the else of this doesn't set out.%s = nil; out could be a pre-existing object...

**Comment by lavalamp** [comment]
So my major concern here is how does this interact with human defined conversion functions? When we make an API change, can I still write a conversion function like I could before? Or another way of putting it, can we turn this on for v1beta1/2?

**Comment by wojtek-t** [comment]
>
So my major concern here is how does this interact with human defined conversion functions? When we make an API change, can I still write a conversion function like I could before? Or another way of putting it, can we turn this on for v1beta1/2?

You definitely should be able to define the conversion function yourself (currently the conversion function wouldn't be generated if it's not obvious how to translate it - i.e. if there are field with different names, or of types that we couldn't automatically detect how to convert). In such cases, writing your own conversion function is obvious (it wouldn't conflict with anything)
The only question is what happens if translation could be detected automatically but we want the conversion to be done in a different way. But I think that we should block this PR on it.

**Comment by lavalamp** [comment]
I see. I was hoping we could read the manually written conversion functions, and use them to generate the expanded conversion functions.

> The only question is what happens if translation could be detected automatically but we want the conversion to be done in a different way. But I think that we should block this PR on it.

I think this effectively will prevent us from mutating our api. We'll need a solution for this soon. I think we can still merge this if you have a plan to fix this soon.

**Comment by wojtek-t** [comment]
>
I see. I was hoping we could read the manually written conversion functions, and use them to generate the expanded conversion functions.

My plan here is to change the manually written conversion functions to be named (with some explicitly defined naming convention) and the use them while generating conversion functions. I just didn't get to it yet.

>
I think this effectively will prevent us from mutating our api. We'll need a solution for this soon. I think we can still merge this if you have a plan to fix this soon.

Yes, I thought about it and I think it should be possible to do it pretty soon. Basically my plan is to:
1. internally, store the generated and manually-written conversion functions separately
2. in default (reflection-based) implementation, check whether any of those exists
3. while generating conversion methods, if there is a manually-written conversion function, do NOT generate one for it
It should be relatively simple PR and should fix the problem.

File: pkg/conversion/generator.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/conversion/generator.go]: Done (here are in few other places).

**Comment by wojtek-t** [comment]
PTAL (hopefully the above explanation will satisfy you :))

**Comment by smarterclayton** [comment]
----- Original Message -----
> >
> I see. I was hoping we could read the manually written conversion functions,
> and use them to generate the expanded conversion functions.
> 
> My plan here is to change the manually written conversion functions to be
> named (with some explicitly defined naming convention) and the use them
> while generating conversion functions. I just didn't get to it yet.
> 
> >
> I think this effectively will prevent us from mutating our api. We'll need a
> solution for this soon. I think we can still merge this if you have a plan
> to fix this soon.
> 
> Yes, I thought about it and I think it should be possible to do it pretty
> soon. Basically my plan is to:
> 1. internally, store the generated and manually-written conversion functions
> separately
> 2. in default (reflection-based) implementation, check whether any of those
> exists
> 3. while generating conversion methods, if there is a manually-written
> conversion function, do NOT generate one for it

Would it be an expectation on the manual-written conversion author to invoke the generated function by name?  I.e.:

func myconversion() {
  out.Bar = in.Bar
  convert_api_SomeObject_to_v1beta3_SomeObject(out, in)
}

?  If so, that seems reasonable.

**Comment by wojtek-t** [comment]
>
Would it be an expectation on the manual-written conversion author to invoke the generated function by name?  I.e.:

func myconversion() {
  out.Bar = in.Bar
  convert_api_SomeObject_to_v1beta3_SomeObject(out, in)
}

?  If so, that seems reasonable.

That's exactly what I thought about.

**Comment by wojtek-t** [comment]
@lavalamp can you please take a look?

**Comment by lavalamp** [comment]
LGTM

Another future improvement would be to inline the defaulting code. It's making a lot of reflect.TypeOf calls right now.

**Comment by wojtek-t** [comment]
>
Another future improvement would be to inline the defaulting code. It's making a lot of reflect.TypeOf calls right now.

Sure - but that's just an optimization.

**Comment by wojtek-t** [comment]
The #7814 showed how important is to have ability to support both manually created and automatically generated conversions together.
That's why I've already written #7832 for support both manually-written and auto-generated conversion functions.
I will rebase this PR only after #7832 is merged (this should be easier to do).

**Comment by wojtek-t** [comment]
I've rebased this PR - the benchmark result are slightly worse now, because auto-generated functions are not explicitly calling manually-written ones - they are called implicitly via Convert() method.
However, the gain is still significant (~8.5X) and we can fix that in the future.

Tthere are no functional changes from the previous version - just regenerated conversion functions + slightly optimized the manually written ones.

PTAL

**Comment by smarterclayton** [comment]
Going to merge

**Comment by wojtek-t** [comment]
Thanks @smarterclayton !

**Comment by lavalamp** [comment]
> we can fix that in the future.

SGTM :)

**Pull request merged**